### PR TITLE
create a local notification authority to ensure processed notificatio…

### PR DESCRIPTION
**Pull request by agrinman**
…ns are created by krypton only

### CODE REVIEW

File: Krypton/LocalNotificationAuthority.swift [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+class LocalNotificationAuthority {
+    
+    internal struct SignedLocalRequest {
+        let localRequestJSON:String
+        let signature:Data
+    }
+    
+    enum Errors:Error {
+        case invalidLocalNotificationSignature
```
Comment (kcking) [inline Krypton/LocalNotificationAuthority.swift]: this case looks unused

File: Krypton/LocalNotificationAuthority.swift [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        
+        guard let signature = KRSodium.instance().auth.tag(message: jsonData, secretKey: signingKey) else {
+            throw Errors.signatureCreationFailed
+        }
+        
+        let localRequest = SignedLocalRequest(localRequestJSON: jsonString, signature: signature)
+        return localRequest.object
+    }
+    
+    static func verifiedLocalNotification(with payload:[String:Any]) throws -> VerifiedLocalRequest {
```
Comment (kcking) [inline Krypton/LocalNotificationAuthority.swift]: I think `verifyLocalNotification` is more clear of a name

File: Krypton/LocalNotificationAuthority.swift [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        guard KRSodium.instance().auth.verify(message: rawRequestData, secretKey: signingKey, tag: rawRequest.signature)
+        else {
+            throw Errors.badLocalRequestSignature
+        }
+        
+        let verifiedLocalRequest = try VerifiedLocalRequest(jsonData: rawRequestData)
+        return verifiedLocalRequest
+    }
+    
+    static func unverifiedLocalNotification(with payload:[String:Any]) throws -> UnverifiedLocalRequest {
```
Comment (kcking) [inline Krypton/LocalNotificationAuthority.swift]: I think `parseUnverifiedLocalNotification` is more clear of a name

File: Krypton/Notify.swift [hunk 4]
```diff
@@ -115,7 +114,14 @@ class Notify {
...
                 content.body = noteBody
                 content.sound = UNNotificationSound.default()
-                content.userInfo = ["session_display": session.pairing.displayName, "session_id": session.id, "request": request.object]
+                
+                let localRequest = LocalNotificationAuthority.VerifiedLocalRequest(alertText: noteBody,
+                                                                                   request: request,
+                                                                                   sessionID: session.id,
+                                                                                   sessionName: session.pairing.displayName)
+                
+                content.userInfo = (try? LocalNotificationAuthority.createSignedPayload(for: localRequest)) ?? [:]
```
Comment (kcking) [inline Krypton/Notify.swift]: should an error here instead be displayed as an error in the notification content?

File: Krypton/Notify.swift [hunk 5]
```diff
@@ -182,6 +186,14 @@ class Notify {
             self.noteMutex.unlock()
             
             log("pushing note with id: \(noteId)")
+            
+            let localRequest = LocalNotificationAuthority.VerifiedLocalRequest(alertText: content.body,
+                                                                               request: request,
+                                                                               sessionID: session.id,
+                                                                               sessionName: session.pairing.displayName)
+            
+            content.userInfo = (try? LocalNotificationAuthority.createSignedPayload(for: localRequest)) ?? [:]
```
Comment (kcking) [inline Krypton/Notify.swift]: should an error here instead be displayed as an error in the notification content?

File: Notify/NotificationService.swift [hunk 6]
```diff
@@ -78,52 +79,60 @@ class NotificationService: UNNotificationServiceExtension {
...
-                            let (noteSubtitle, noteBody) = unsealedRequest.notificationDetails()
-                            
-                            content.subtitle = noteSubtitle
-                            content.body = noteBody
-
-                            // special case: me request
-                            if case .me = unsealedRequest.body {
-                                content.title = "\(session.pairing.displayName)."
-                            }
+                        self.bestAttemptMutex.lock()
```
Comment (kcking) [inline Notify/NotificationService.swift]: why the switch from `.lock{ }` to `.lock()`?

File: Krypton/LocalNotificationAuthority.swift (hunk 1 above)
Comment (agrinman) [reply Krypton/LocalNotificationAuthority.swift]: 👍

File: Krypton/LocalNotificationAuthority.swift (hunk 2 above)
Comment (agrinman) [reply Krypton/LocalNotificationAuthority.swift]: 👍

File: Krypton/LocalNotificationAuthority.swift (hunk 3 above)
Comment (agrinman) [reply Krypton/LocalNotificationAuthority.swift]: 👍

File: Krypton/Notify.swift (hunk 4 above)
Comment (agrinman) [reply Krypton/Notify.swift]: 👍

File: Krypton/Notify.swift (hunk 5 above)
Comment (agrinman) [reply Krypton/Notify.swift]: 👍

File: Notify/NotificationService.swift (hunk 6 above)
Comment (agrinman) [reply Notify/NotificationService.swift]: rewrote mutex logic, check the latest commit

**Pull request merged**